src/collapse/xdebug.rs

```diff
@@ -0,0 +1,367 @@
+}
+
+struct Frames<'a> {
+    calls: &'a CallStack,
+    stack: &'a [usize],
+}
+
+/// Iterate over tab separated fields.
+///
+/// This is essentially the same as `str.split('\t').filter(|slice| slice.len() > 0)` but seemingly
+/// faster. This probably because we know that the character in question is ascii, while the
+/// standard library `StrSearcher` for `char` is not optimized around this case.
+struct Fields<'a> {
+    line: &'a str,
+}
+
+/// The Folder struct
+pub struct Folder;
+
+impl From<Options> for Folder {
+    fn from(_: Options) -> Self {
+        Self
+    }
+}
+
+impl Collapse for Folder {
+    fn is_applicable(&mut self, input: &str) -> Option<bool> {
+        let mut input = input.as_bytes();
+        let mut line = String::new();
+        loop {
+            line.clear();
+
+            if let Ok(n) = input.read_line(&mut line) {
+                if n == 0 {
+                    break;
+                } else {
+                    return Some(false);
+                }
+            }
+
+            if line.starts_with(TRACE_START) {
+                return Some(true);
+            }
+        }
+
+        None
+    }
+
+    fn collapse<R, W>(&mut self, mut reader: R, mut writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        // Timings for all call stacks in total.
+        let mut stacks: HashMap<_, f32> = HashMap::new();
+        let mut current_stack = CallStack::new();
+        let mut prev_start_time = 0.0;
+        let mut line = String::new();
+
+        loop {
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            if line.starts_with(TRACE_START) {
+                break;
+            }
+
+            line.clear();
+        }
+
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            if line.starts_with(TRACE_END) {
+                break;
+            }
+
+            // Break the line into tab separated tokens.
+            let mut parts = Fields::new(&line).skip(2);
+
+            let (is_exit, time) = if let (Some(is_exit), Some(time)) = (parts.next(), parts.next()) {
+                let is_exit = match is_exit {
+                    "1" => true,
+                    "0" => false,
+                    a => panic!(format!("uh oh: {}", a)),
+                };
+
+                let time = time.parse::<f32>().unwrap();
+
+                (is_exit, time)
+            } else {
+                continue;
+            };
+
+            if is_exit && current_stack.is_empty() {
+                warn!("Found function exit without corresponding entrance. Discarding line. Check your input.");
+                continue;
+            }
+
+            {
+                let current = current_stack.current();
+                let duration = SCALE_FACTOR * (time - prev_start_time);
+
+                // hash `current` for lookup and insertion.
```

Comment:
I think it'd be useful to add a comment here about exactly why you're doing this in such a roundabout way. I _suspect_ it is because you don't want to clone the string if you don't have to do the insertion, which the regular entry API would require you to do?

@jasonrhansen this might be something we want to do for the other collapse implementations too.

Reply:
One problem is we just switched back to using `HashMap` in `std`, but the raw entry API is nightly-only there. We'd have to switch back to `hashbrown`.

Reply:
Ah, good point, I didn't think of that... I think I'd be okay with switching back if we found this to yield a significant speed-up though (and it may well do!).